sound/soc/sof/pm.c

```diff
@@ -138,25 +146,38 @@ static int sof_restore_pipelines(struct snd_sof_dev *sdev)
 		}
 	}
 
-	/* restore dai links */
+	link_array = devm_kzalloc(sdev->dev, sizeof(int) * (max_index + 1),
+				  GFP_KERNEL);
+	if (!link_array)
+		return -ENOMEM;
+
+	/* restore dai links from index 0 to max index */
 	list_for_each_entry_reverse(dai, &sdev->dai_list, list) {
 		struct sof_ipc_reply reply;
 		struct sof_ipc_dai_config *config = dai->dai_config;
 
-		ret = sof_ipc_tx_message(sdev->ipc,
-					 config->hdr.cmd, config,
-					 config->hdr.size,
-					 &reply, sizeof(reply));
-
-		if (ret < 0) {
-			dev_err(sdev->dev,
-				"error: failed to set dai config for %s\n",
-				dai->name);
+		if (!link_array[config->dai_index]) {
```

Comment:
@RanderWang i'm a bit confused by this logic because there are 3 types of dai's, each with indices ranging from 0 - max_of_each_type. But you use only one array of 0-max. So depending on which order the dai's were added, you could miss out on configuring some dai's. please correct me if i am wrong.

Also, we dont really have any issues because of sendign the dai config twice for the same dai. Is it really important to avoid the second IPC if it does no harm?

Reply:
> @RanderWang i'm a bit confused by this logic because there are 3 types of dai's, each with indices ranging from 0 - max_of_each_type. But you use only one array of 0-max. So depending on which order the dai's were added, you could miss out on configuring some dai's. please correct me if i am wrong.
yes, there maybe 3 types of dais but not each indices ranging from 0 - max. All of them are in the range of 0 - max

> Also, we dont really have any issues because of sending the dai config twice for the same dai. Is it really important to avoid the second IPC if it does no harm?
no, we have a issue. If one dai config is set two times, one DMA channel would be wasted in FW. please refer to https://github.com/thesofproject/sof/pull/786

Reply:
@RanderWang thanks for confirming that this is required. 

But about my first comment, so can different types of DAI's not have the same dai index?

Reply:
looking at sof-apl-nocodec.m4 , i feel like the assumption that the dai index for all DAI's is unique is wrong.

Reply:
I check the file and don't find unique id.  First we should make clear the usage of dai index. It is used in ASoC which doesn't care the type of dai. It just identifies dai by index.

Reply:
@ranj063  I get your idea. I think it should be wrong in topology. I will check FW  and discuss it with keyon. Thanks!

Reply:
Yes,  one choice is to let FW make it.  And my idea is to maintain a dai link list in sof like dai list. With  dai link list, everything is simple and easy to understand

Reply:
Meh. if the firmware can detect the configs are the same, why couldn't the driver do it as well?

Reply:
@plbossart thanks for your comment. It make easier for me now. And the PR in FW is ready thesofproject/sof#786

Reply:
What does this FW PR have to do with configs being sent twice?

Reply:
dma channel is not released when config is set again. This PR would check whether dma channel is allocated for a dai link config and avoid allocating dma channel again @keyonjie

Reply:
@RanderWang so are you suggesting we just send the config twice as the FW can handle it now?